src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -2552,16 +2475,11 @@ public void EndDisconnect(IAsyncResult asyncResult)
                 throw new ObjectDisposedException(this.GetType().FullName);
             }
 
-#if FEATURE_PAL
```

Comment:
That if we have a valid AsyncResult, we must be on Windows.

Looking at the code a bit more, I'm not sure we can actually assert that, because we only check that it's a LazyAsyncResult, and it's entirely possible to get one of these somewhere else.  That seems like a different problem, though.